Introduce Metal3 plugin for openshift console

jtomasek opened the pull request:
This is the initial introduction of the metal3 plugin. Currently or testing purposes only.

~~Depends on https://github.com/openshift/console/pull/1586~~

TODO:
- [x] ~~remove web-ui-components dependency~~
- [x] split into multiple PRs
- [x] add OWNERS file

jtomasek commented on frontend/public/kinds.ts:
```diff
@@ -35,8 +34,8 @@ export const connectToPlural: ConnectToPlural = connect((state: State, props: {p
-    ? state.k8s.getIn(['RESOURCES', 'models']).get(plural)
-    : _.find(k8sModels, model => model.plural === plural);
+    ? state.k8s.getIn(['RESOURCES', 'models', plural])
+    : state.k8s.getIn(['RESOURCES', 'models']).find(model => model.plural === plural);
```
This allows the models to be defined in the plugin rather than in common k8sModels module

spadgett replied:
I think this will break some other cases. For instance, if there are multiple `deployments` in different API versions, we might get the wrong one. We need to make sure we get the one defined in the static model.

vojtechszocs replied:
I agree that Console should keep referring to existing model definitions (`public/models/index.ts`) while **also** allowing plugins to contribute new model definitions.

I think the best way to solve this is via auto-generated `@console/plugin-models` module which exports additional model definitions for Console application to work with.

@jelkosz ^^ please create "Allow plugins to contribute new model definitions" task & assign it to me.

jelkosz replied:
@vojtechszocs ok, done

vojtechszocs replied:
> I think the best way to solve this is via auto-generated `@console/plugin-models` module which exports additional model definitions for Console application to work with.

I've re-thinked this and will avoid generating another virtual module in favor of collecting models from extension objects at runtime and augmenting `public/models/index.ts` module exports.

vojtechszocs replied:
#1586 should fix this.

spadgett commented on frontend/packages/console-metal3/src/components/host.tsx:
```diff
@@ -0,0 +1,150 @@
+  namespace: string;
+};
+
+export class BaremetalHostsPage extends React.Component<
```
Use a functional component instead of a class

jtomasek replied:
Done

spadgett commented on frontend/packages/console-metal3/src/components/host.tsx:
```diff
@@ -0,0 +1,150 @@
+        />
+      </div>
+      <div className={statusColumnClasses}>
+        <WithResources resourceMap={machineName ? hostResourceMap : {}}>
```
We should not be opening WebSockets for individual table cells

vojtechszocs replied:
I remember addressing this kind of issue in our Console fork, @rawagner can you please share the details?

mareklibra replied:
We dealt with this on multiple places, i.e. to show VM status which depends on multiple referenced objects. Optimized by requesting larger dataset via lower count of WebSockets and filtering on UI.

spadgett commented on frontend/packages/console-shared/src/components/utils/with-resources.tsx:
```diff
@@ -0,0 +1,167 @@
+};
+
+/*
+ * Firehose helper
```
Can you explain what this does and why it's needed?

vojtechszocs replied:
This seems to be a KubeVirt utility, ping @rawagner @suomiy

https://github.com/kubevirt/web-ui/blob/master/frontend/public/kubevirt/components/utils/withResources.jsx

rawagner replied:
Yes, it is KubeVirt utility which simplifies work with Firehose and adds some features.
For example:
- `Firehose` wont render children if model does not exist in k8s. `WithResources` will - both cases has their place
- `Resources` accepts resourceMap which is object instend of array as in Firehose which brings us a few enhancements:

  - you need to define key (which is used as firehose resource prop - just makes it easier if you have more queries for the same model, object forces you to use unique keys)

  - you can define if resource from resourceMap is required - if yes we wait with child rendering until the resource is loaded (or we can show Loading)
- there is also way to define your own error handling
- you no longer need to check firehose's loaded prop - when injected resource is undefined it means it is still loading

vojtechszocs commented on frontend/packages/console-app/package.json:
```diff
@@ -7,6 +7,7 @@
   "dependencies": {
     "@console/internal": "0.0.0-fixed",
     "@console/plugin-sdk": "0.0.0-fixed",
+    "@console/metal3": "0.0.0-fixed",
```
Committing this change to git means that Metal 3 plugin will be included in development and production builds by default.

Until the Metal 3 plugin leaves the experimental phase, I'd suggest to treat this as a local change and not commit to git.

jtomasek replied:
Done

vojtechszocs commented on frontend/packages/console-metal3/package.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "@console/metal3",
```
I'd suggest to follow these conventions:

- package path
  - should include the `plugin` suffix to denote the purpose of the package
  - should not include the `console` prefix (`console-xxx` is reserved for core Console packages)
  - proposed value: `packages/metal3-plugin`

- package name
  - `@console` scope is required due to monorepo and tooling
  - proposed value: `@console/metal3-plugin`

Later on, Metal3 team could add & own other packages as necessary, e.g. `packages/metal3-components` etc.

jtomasek replied:
Done

vojtechszocs commented on frontend/packages/console-metal3/src/components/host.tsx:
```diff
@@ -0,0 +1,150 @@
+  getResource,
+  getSimpleHostStatus,
+  getUid
+} from 'kubevirt-web-ui-components';
```
We need to add this dependency into the appropriate `package.json` file.

For now, all **core** Console dependencies are declared in the monorepo root package. These dependencies reflect the requirements of application code within the `public` directory, to be moved into the `@console/app` package in future.

As for `kubevirt-web-ui-components`, this is a plugin-specific dependency so please add it to Metal3 plugin's `package.json` file.

jtomasek replied:
Done

vojtechszocs commented on frontend/packages/console-metal3/src/components/machine-cell.jsx:
```diff
@@ -0,0 +1,49 @@
+import { Icon } from 'patternfly-react';
+import {
+  canHostAddMachine,
+  DASHES,
```
Nit: isn't there some constant that we can import from `@console/internal` instead?

jtomasek replied:
It seems that simple '-' string is used frequently in the codebase.

vojtechszocs commented on frontend/packages/console-metal3/src/models.ts:
```diff
@@ -0,0 +1,15 @@
+  namespaced: true,
+  kind: 'BareMetalHost',
+  id: 'baremetalhost',
+  crd: true
```
Nit: missing trailing comma, you can run `yarn lint` in the monorepo root to ensure there are no linting issues.

jtomasek replied:
Done

vojtechszocs commented on frontend/packages/console-metal3/src/plugin.ts:
```diff
@@ -0,0 +1,47 @@
+  {
+    type: 'NavItem/ResourceNS',
+    properties: {
+      section: 'Compute',
```
This currently yields a TypeScript error since `ExtensionProperties.NavItem.section` is hard-coded as:

```
'Home' | 'Workloads'
```

I'm working on a PR to fix this.

In the meantime, please update the `packages/console-plugin-sdk/src/typings/nav.ts` module to include the sections you need.

vojtechszocs replied:
cc @jelkosz ^^ is related to the "Improve nav item extension capabilities" task.

jtomasek replied:
Done

vojtechszocs replied:
#1584 should fix this.

vojtechszocs commented on frontend/packages/console-metal3/src/plugin.ts:
```diff
@@ -0,0 +1,47 @@
+      model: BaremetalHostModel,
+      loader: () =>
+        import(
+          './components/host' /* webpackChunkName: "baremetalhost" */
```
The `webpackChunkName` value should reflect the name of your plugin for better traceability.

I'd suggest to start with a single split point, i.e. use `webpackChunkName: "metal3"` for all async module requests.

spadgett replied:
We typically use different chunk names for different pages for better code splitting. Then the JS is only loaded when you go to that page.

vojtechszocs replied:
OK, sounds reasonable :smiley:

Looking at `public/components/resource-pages.ts` the convention seems to be to put both resource list and detail pages under the same split point, e.g. `pod` for both `PodsPage` and `PodsDetailsPage` (group by resource).

jtomasek replied:
I've used `metal3-baremetalhost`

vojtechszocs commented on frontend/packages/console-shared/src/components/utils/with-resources.tsx:
```diff
@@ -0,0 +1,167 @@
+import * as React from 'react';
```
There doesn't seem to be a corresponding code removal in context of the `public` directory, where does this code originate from?

@rawagner Is this code used in `kubevirt/web-ui` fork?

jtomasek replied:
Right, the code is coming from kubevirt/web-ui, I had to add it as hosts listing depends on it. I am happy to split it into a separate PR and update this one to depend on it if needed.

vojtechszocs commented:
FYI, the `kubevirt-web-ui-components` dependency also implies adding `@patternfly/react-charts`.

spadgett commented:
> FYI, the `kubevirt-web-ui-components` dependency also implies adding `@patternfly/react-charts`.

OK, we're doing that anyway in #1448

vojtechszocs commented:
> > FYI, the `kubevirt-web-ui-components` dependency also implies adding `@patternfly/react-charts`.
> 
> OK, we're doing that anyway in #1448

Thanks for the update.

#1448 adds `"@patternfly/react-charts": "^3.4.2"`, however the latest `kubevirt-web-ui-components` release has a peer dependency on `"@patternfly/react-charts": "2.x"` which should be bumped.

@rawagner @mareklibra ^^ we should fix outdated peer dependencies in `kubevirt-web-ui-components`.

spadgett commented:
> #1448 adds `"@patternfly/react-charts": "^3.4.2"`, however the latest `kubevirt-web-ui-components` release has a peer dependency on `"@patternfly/react-charts": "2.x"` which should be bumped.
> 
> @rawagner @mareklibra ^^ we should fix outdated peer dependencies in `kubevirt-web-ui-components`.

FYI, #1448 merged and now has `"@patternfly/react-charts": "^3.4.5"`

vojtechszocs commented:
@jtomasek Please rebase on top of #1586 (includes commits from #1528 and #1584).

vojtechszocs commented:
@jtomasek FYI, if you need to add YAML templates for your models, check out #1613.

jtomasek commented:
Rebased on top of https://github.com/openshift/console/pull/1586
Simplified to avoid using WithResources which will be added in a subsequent changes
Added OWNERS file

jtomasek commented on frontend/packages/metal3-plugin/src/plugin.ts:
```diff
@@ -0,0 +1,54 @@
+    type: 'FeatureFlag/Model',
+    properties: {
+      model: BaremetalHostModel,
+      flag: 'METAL3',
```
Done

vojtechszocs commented:
> I am curious to see how often we will be getting merge conflicts in `yarn.lock` in practise.

There's one `yarn.lock` per the whole Console frontend monorepo, so we'll have to deal with those merge conflicts. Instead of manually sync'ing upstream `yarn.lock` vs. your local `yarn.lock`, you can rebase and execute dependency-mgmt Yarn commands on top of the latest upstream `yarn.lock`.

vojtechszocs commented on frontend/packages/metal3-plugin/src/plugin.ts:
```diff
@@ -0,0 +1,56 @@
+  {
+    type: 'ModelDefinition',
+    properties: {
+      models: [BaremetalHostModel],
```
You can also do

```diff
-import { BaremetalHostModel } from './models';
+import * as models from './models';
```

```diff
-models: [BaremetalHostModel],
+models: _.values(models),
```

like in #1592 to avoid manually importing individual models (if you expect to have lots of them).

jtomasek replied:
I don't expect there to be many more models, so I listed it explicitly here

vojtechszocs replied:
OK, I'm fine with that.

vojtechszocs commented on frontend/packages/metal3-plugin/src/plugin.ts:
```diff
@@ -0,0 +1,56 @@
+        name: 'Bare Metal Hosts',
+        resource: 'baremetalhosts',
+        required: METAL3_FLAG
+      },
```
If you want to put your nav link after an existing one, you can use `mergeAfter`.

jtomasek replied:
Done

spadgett commented:
```
/go/src/github.com/openshift/console/frontend/packages/metal3-plugin/src/plugin.ts
  19:29  error  Missing semicolon       semi
  32:24  error  Missing trailing comma  comma-dangle
  42:30  error  Missing trailing comma  comma-dangle

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the `--fix` option.
```

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+  // getHostMachineName,
+  getHostBmcAddress,
+  getName,
+  getNamespace,
```
Same comment as in #1592, these aren't kubevirt specific. They should be added somewhere that can be used generally across console.

jtomasek replied:
I am happy to move this to console/shared in a subsequent PR and gradually reduce the dependency on web-ui-components. To unblock work depending on this PR. Would that be sufficient?

mareklibra replied:
We plan to move `kubevirt-web-ui-components` to `console/shared` or similar packages within one of the last steps of merging `web-ui`. Until this happens, we would need to maintain these utility functions (or similar code) on two places - `kubevirt-web-ui` is still in play until `openshift-console` meets full feature parity.

See my response in #1592 for more details.

vojtechszocs replied:
> I am happy to move this to console/shared in a subsequent PR and gradually reduce the dependency on web-ui-components.

The idea was that `@console/shared` contains code used by Console itself as well as other packages. The typical contribution workflow here would be moving code from `public` or `@console/app` to `@console/shared`. Package owners should be inherited from monorepo root `OWNERS`.

For code not needed by Console itself, we could add a dedicated package, e.g. `@console/rhhi-shared` (or similar). Common `kubevirt/web-ui` or `kubevirt-web-ui-components` code (KubeVirt, Metal3, Storage) could be moved there. Package owners should be declared explicitly, including people who work on said projects.

> We plan to move `kubevirt-web-ui-components` to `console/shared` or similar packages within one of the last steps of merging `web-ui`.

I'd create a dedicated package for that, e.g. `@console/rhhi-shared` suggested above.

(`kubevirt-web-ui-components` is not KubeVirt only, it also covers Metal3 and Storage.)

jtomasek replied:
I agree that for getting rid of kubevirt/web-ui-components dependency creating of a separate console/rhhi-shared package makes sense as it would contribute to speeding up the process.

On the other hand in ideal world it seems to me that package specific code can live in the package directly (e.g. metal3 specific components) and common ones should align to general use in console and therefore end up in console/shared. But I agree that we're not there yet and it will take time to get there, so the solution you propose is perfectly reasonable.

Regarding Sam's specific comment, these selectors can be used across the whole console codebase so I think it is quite reasonable to put it directly to console/shared.

vojtechszocs replied:
> package specific code can live in the package directly (e.g. metal3 specific components) and common ones should align to general use in console and therefore end up in console/shared

Yes, that's an ideal solution - `console-shared` changes should pass core team reviews, so any new code we pull from `kubevirt/web-ui` fork or `kubevirt-web-ui-components` library would be revisited and aligned with Console core.

We should aim for the best possible solution, so +1 on just going straight through `console-shared`, bit by bit.

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+const addressColumnClasses = 'col-lg-2 visible-lg';
+
+const HostHeader: React.FC<React.ComponentProps<typeof ColHead>> = props => (
+  <ListHeader>
```
Heads up that we're in the process of updating all tables to PF4 (#1465)

@priley86 FYI

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+  return (
+    <ResourceRow obj={host}>
+      <div className={nameColumnClasses}>
+        <ResourceLink kind={BaremetalHostModel.kind} name={name} namespace={namespace} title={uid} />
```
I'd remove `title={uid}`

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+        <ResourceLink kind={BaremetalHostModel.kind} name={name} namespace={namespace} title={uid} />
+      </div>
+      <div className={statusColumnClasses}>
+        {/* <WithResources resourceMap={machineName ? hostResourceMap : {}}>
```
I'd remove the empty table columns and add them back when we're ready. We try to keep master as close to production quality as we can.

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+//   },
+// ];
+
+/* eslint-disable no-unused-vars, no-undef */
```
You should be able to remove this comment.

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,130 @@
+// ];
+
+/* eslint-disable no-unused-vars, no-undef */
+export type BaremetalHostsPageProps = {
```
Unnecessary export

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/machine-cell.tsx:
```diff
@@ -0,0 +1,45 @@
+        title={machineName}
+      />
+    );
+  } else if (canHostAddMachine(host)) {
```
Unnecessary `else` since you return inside the `if`

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/machine-cell.tsx:
```diff
@@ -0,0 +1,45 @@
+    const ref = referenceForModel(MachineModel);
+    const href = `/k8s/ns/${namespace || 'default'}/${ref}/new`;
+    return (
+      <Link to={href}>
```
Use `RequireCreatePermission` or another RBAC check.

jtomasek replied:
Done

jtomasek commented on frontend/packages/metal3-plugin/OWNERS:
```diff
@@ -0,0 +1,10 @@
+approvers:
+  - jtomasek
+  - honza
+  - flofuchs
```
Done

spadgett replied:
> Done

Looks like it's still missing the final newline

spadgett commented:
/retest

spadgett commented on frontend/packages/metal3-plugin/OWNERS:
```diff
@@ -0,0 +1,11 @@
+  - jtomasek
+  - honza
+  - flofuchs
+  
```
nit: extra newline

jtomasek replied:
ugh, @vojtechszocs suggested adding it in the last comment :) https://github.com/openshift/console/pull/1539#discussion_r289427223

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,125 @@
+// const statusColumnClasses = 'col-lg-2 col-md-4 hidden-sm hidden-xs';
+const machineColumnClasses = 'col-lg-3 visible-lg';
+// const roleColumnClasses = 'col-lg-2 visible-lg';
+const addressColumnClasses = 'col-lg-2 visible-lg';
```
I know this temporary, but columns no longer add up to 12. You might add a column like created to fill things out until we are able to add the others.

jtomasek replied:
I've added common column class for the 3 current columns and will adjust it back once remaining columns are added.

spadgett commented on frontend/packages/metal3-plugin/src/selectors/index.ts:
```diff
@@ -0,0 +1,15 @@
+export const isHostOnline = host => _.get(host, 'spec.online', false);
+export const getHostNics = host => _.get(host, 'status.hardware.nics', []);
+export const getHostStorage = host => _.get(host, 'status.hardware.storage', []);
+export const getHostCpus = host => _.get(host, 'status.hardware.cpus', []);
```
```suggestion
export const getHostCPUs = host => _.get(host, 'status.hardware.cpus', []);
```

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/selectors/index.ts:
```diff
@@ -0,0 +1,15 @@
+export const getHostMachineName = host => _.get(host, 'spec.machineRef.name');
+export const getHostBmcAddress = host => _.get(host, 'spec.bmc.address');
+export const isHostOnline = host => _.get(host, 'spec.online', false);
+export const getHostNics = host => _.get(host, 'status.hardware.nics', []);
```
```suggestion
export const getHostNICs = host => _.get(host, 'status.hardware.nics', []);
```

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/selectors/index.ts:
```diff
@@ -0,0 +1,15 @@
+export const getOperationalStatus = host => _.get(host, 'status.operationalStatus');
+export const getProvisioningState = host => _.get(host, 'status.provisioning.state');
+export const getHostMachineName = host => _.get(host, 'spec.machineRef.name');
+export const getHostBmcAddress = host => _.get(host, 'spec.bmc.address');
```
```suggestion
export const getHostBMCAddress = host => _.get(host, 'spec.bmc.address');
```

jtomasek replied:
Done

spadgett commented on frontend/packages/metal3-plugin/src/selectors/index.ts:
```diff
@@ -0,0 +1,15 @@
+export const getHostNics = host => _.get(host, 'status.hardware.nics', []);
+export const getHostStorage = host => _.get(host, 'status.hardware.storage', []);
+export const getHostCpus = host => _.get(host, 'status.hardware.cpus', []);
+export const getHostRam = host => _.get(host, 'status.hardware.ramGiB');
```
```suggestion
export const getHostRAM = host => _.get(host, 'status.hardware.ramGiB');
```

jtomasek replied:
Done

jtomasek commented:
/retest

jtomasek commented on frontend/packages/console-app/package.json:
```diff
@@ -7,7 +7,8 @@
     "@console/plugin-sdk": "0.0.0-fixed",
-    "@console/shared": "0.0.0-fixed"
+    "@console/shared": "0.0.0-fixed",
+    "@console/metal3-plugin": "0.0.0-fixed"
```
This needs to be removed

jtomasek replied:
Done

jtomasek commented on frontend/packages/metal3-plugin/src/components/host.tsx:
```diff
@@ -0,0 +1,133 @@
+  </ListHeader>
+);
+
+const HostRow = ({ obj: host }: React.ComponentProps<typeof ResourceRow>) => {
```
I had to switch from using React.FC as it did not satisfy react/prop-types lint rule.
(getting `error  'obj' is missing in props validation  react/prop-types`)

vojtechszocs commented on frontend/packages/metal3-plugin/src/models.ts:
```diff
@@ -0,0 +1,14 @@
+import { K8sKind } from '@console/internal/module/k8s';
+
+export const BaremetalHostModel: K8sKind = {
```
See my [comment on KubeVirt models](https://github.com/openshift/console/pull/1592#discussion_r290353110) - shouldn't we use `crd: true` here?

@spadgett In `kubevirt/web-ui` fork, any additional models are declared _without_ the `crd` flag, which is a mistake IMO.

jtomasek replied:
I agree, although when I add the crd flag, I am getting `kindObj: no model for kind BareMetalHost` in the console and the ListPage won't show up. It seems like there is a mismatch in keys of the resources in redux store and how crd resources are retrieved.

jtomasek replied:
Fixed the problem and added `crd: true` flag to the model.

vojtechszocs replied:
> It seems like there is a mismatch in keys of the resources in redux store and how crd resources are retrieved.

There's currently a duality in how Console references models. Non-CRD models are typically referenced only by their `kind`, CRD models are referenced via `referenceForModel` function that takes their `apiGroup`, `apiVersion` and `kind` into account.

Getting `kindObj: no model for kind BareMetalHost` could mean you aren't referencing your CRD model via `referenceForModel` function.

vojtechszocs commented:
@jtomasek #1668 changes resource page extension types, please adjust accordingly:
- `ResourcePage/List` => `Page/Resource/List`
- `ResourcePage/Detail` => `Page/Resource/Details`

jtomasek commented:
> @jtomasek #1668 changes resource page extension types, please adjust accordingly:
> 
> * `ResourcePage/List` => `Page/Resource/List`
> * `ResourcePage/Detail` => `Page/Resource/Details`

Done

jtomasek commented:
/retest

spadgett commented on frontend/packages/metal3-plugin/src/components/machine-cell.tsx:
```diff
@@ -0,0 +1,48 @@
+    const ref = referenceForModel(MachineModel);
+    const href = `/k8s/ns/${namespace || 'default'}/${ref}/new`;
+    return (
+      <RequireCreatePermission model={MachineModel} namespace={namespace}>
```
This doesn't handle the all namespaces case correctly

jtomasek replied:
Done

spadgett commented on frontend/yarn.lock:
```diff
@@ -8021,6 +8085,21 @@ kind-of@^6.0.0, kind-of@^6.0.2:
   version "6.0.2"
   resolved "https://registry.yarnpkg.com/kind-of/-/kind-of-6.0.2.tgz#01146b36a6218e64e58f3a8d66de5d7fc6f6d051"
 
+kubevirt-web-ui-components@^0.1.34-2:
```
I'm concerned with all the additional dependencies this is bringing in. It looks like it includes some alternative libraries to things we've already included in console. Is this only being used for the `getName` / `getNamespace` / `getUid` helpers?

I realize you use it in the kubevirt fork for other things.

vojtechszocs replied:
As an alternative, we could move the code we need from `kubevirt-web-ui-components` to `console-shared` (or create another, more focused shared package) and align it with current Console dependencies, code (like `Firehose` vs. `WithResources`) etc.

jtomasek replied:
> Is this only being used for the `getName` / `getNamespace` / `getUid` helpers?

+ a couple of things in machine-cell.tsx

I can quite easily remove this dependency for metal3 plugin in general as it does not have large amount of parts implemented in web-ui-components. For kubevirt plugin It will be more difficult.

jtomasek replied:
With recent change I've removed the kubevirt/web-ui-components dependency and added required selectors to @console/shared package.

vojtechszocs commented on frontend/packages/metal3-plugin/package.json:
```diff
@@ -0,0 +1,14 @@
+  "dependencies": {
+    "@console/plugin-sdk": "0.0.0-fixed",
+    "@console/shared": "0.0.0-fixed",
+    "kubevirt-web-ui-components": "^0.1.34-2"
```
Note: you can evaluate the impact of adding `kubevirt-web-ui-components` on webpack-generated JS dist code by running `yarn analyze` which runs the [webpack-bundle-analyzer](https://github.com/webpack-contrib/webpack-bundle-analyzer).

jtomasek commented:
Removed `kubevirt/web-ui-components` dependency.

spadgett commented:
/retest
